### PR TITLE
VMware: new module: vmware_host_powerstate

### PR DESCRIPTION
**Pull request by Akasurde**
##### SUMMARY
Signed-off-by: Abhijeet Kasurde <akasurde@redhat.com>

##### ISSUE TYPE
 - New Module Pull Request

##### COMPONENT NAME
lib/ansible/modules/cloud/vmware/vmware_host_powerstate.py
test/integration/targets/vmware_host_powerstate/aliases
test/integration/targets/vmware_host_powerstate/tasks/main.yml

##### ANSIBLE VERSION

```
2.6devel
```

### CODE REVIEW

**Comment by Akasurde** [comment]
@pdellaert @tchernomax Could you please review this ?

File: lib/ansible/modules/cloud/vmware/vmware_host_powerstate.py [hunk 1]
```diff
@@ -0,0 +1,222 @@
...
+    - This is required parameter if C(esxi_hostname) is not specified.
+  force:
+    description:
+    - 'This parameter specify if the host should be proceeding with user defined powerstate
+      regardless of whether it is in maintenance mode.'
+    - 'If C(state) set to C(reboot-host) and C(force) as C(true), then host system is rebooted regardless of whether it is in maintenance mode.'
+    - 'If C(state) set to C(shutdown-host) and C(force) as C(true), then host system is shutdown regardless of whether it is in maintenance mode.'
+    - 'If C(state) set to C(power-down-to-standby) and C(force) to C(true), then all powered off VMs will evacuated.'
+    - 'Not applicable if C(state) set to C(power-up-from-standby).'
+    choices: [ True, False ]
```
Comment (pdellaert) [inline lib/ansible/modules/cloud/vmware/vmware_host_powerstate.py]: If type is bool, does it make sense to mention choices?

File: lib/ansible/modules/cloud/vmware/vmware_host_powerstate.py [hunk 2]
```diff
@@ -0,0 +1,222 @@
...
+
+
+def main():
+    argument_spec = vmware_argument_spec()
+    argument_spec.update(
+        state=dict(type='str', default='shutdown-host',
+                   choices=['power-down-to-standby', 'power-up-from-standby', 'shutdown-host', 'reboot-host']),
+        esxi_hostname=dict(type='str', required=False),
+        cluster_name=dict(type='str', required=False),
+        force=dict(type='bool', default=False, choices=[True, False]),
```
Comment (pdellaert) [inline lib/ansible/modules/cloud/vmware/vmware_host_powerstate.py]: Does it make sense to mention choices?

File: lib/ansible/modules/cloud/vmware/vmware_host_powerstate.py [hunk 3]
```diff
@@ -0,0 +1,222 @@
...
+options:
+  state:
+    description:
+    - Set the state of the host system.
+    choices: [ 'power-down-to-standby', 'power-up-from-standby', 'shutdown-host', 'reboot-host' ]
+    default: 'shutdown-host'
+  esxi_hostname:
+    description:
+    - Name of the host system to work with.
+    - This is required parameter if C(cluster_name) is not specified.
```
Comment (pdellaert) [inline lib/ansible/modules/cloud/vmware/vmware_host_powerstate.py:42]: Not added in module definition, (required_one_of)

File: lib/ansible/modules/cloud/vmware/vmware_host_powerstate.py [hunk 4]
```diff
@@ -0,0 +1,222 @@
...
+
+        """
+        results = dict(changed=False, result=dict())
+        state = self.params.get('state')
+        force = self.params.get('force')
+        timeout = self.params.get('timeout')
+        host_change_list = []
+        for host in self.hosts:
+            changed = False
+            if not host.runtime.inMaintenanceMode and not force:
```
Comment (pdellaert) [inline lib/ansible/modules/cloud/vmware/vmware_host_powerstate.py:146]: Does this work if the host is powered down without being in maintenance mode (for instance in standby forced through this module) and you try to power it up from standby?

File: lib/ansible/modules/cloud/vmware/vmware_host_powerstate.py [hunk 5]
```diff
@@ -0,0 +1,222 @@
...
+                self.module.fail_json(msg="Current host system '%s' is not in maintenance mode,"
+                                          " please specify 'force' as True to proceed." % host.name)
+            if host.runtime.connectionState == 'notResponding':
+                self.module.fail_json(msg="Current host system '%s' can not be set in '%s'"
+                                          " mode as the host system is not responding." % (host.name, state))
+
+            results['result'][host.name] = dict(msg='', error='')
+            if state == 'reboot-host' and not host.capability.rebootSupported:
+                self.module.fail_json(msg="Current host '%s' can be rebooted as the host system"
```
Comment (pdellaert) [inline lib/ansible/modules/cloud/vmware/vmware_host_powerstate.py]: `Current host '%s' can not be rebooted...`

File: lib/ansible/modules/cloud/vmware/vmware_host_powerstate.py [hunk 6]
```diff
@@ -0,0 +1,222 @@
...
+                                          " does not have capability to reboot." % host.name)
+            elif state == 'shutdown-host' and not host.capability.shutdownSupported:
+                self.module.fail_json(msg="Current host '%s' can be shut down as the host system"
```
Comment (pdellaert) [inline lib/ansible/modules/cloud/vmware/vmware_host_powerstate.py]: `Current host '%s' can not be shut down...`

File: lib/ansible/modules/cloud/vmware/vmware_host_powerstate.py [hunk 7]
```diff
@@ -0,0 +1,222 @@
...
+                                          " does not have capability to shut down." % host.name)
+            elif state in ['power-down-to-standby', 'power-up-from-standby'] and not host.capability.standbySupported:
+                self.module.fail_json(msg="Current host '%s' can be %s as the host system"
```
Comment (pdellaert) [inline lib/ansible/modules/cloud/vmware/vmware_host_powerstate.py]: `Current host '%s' can not be %s...`

File: lib/ansible/modules/cloud/vmware/vmware_host_powerstate.py [hunk 8]
```diff
@@ -0,0 +1,222 @@
...
+                                          " does not have capability to standby supported." % host.name)
+
+            verb = "reboot '%s'" % host.name
```
Comment (pdellaert) [inline lib/ansible/modules/cloud/vmware/vmware_host_powerstate.py]: shouldn't this be in the `if state == 'reboot-host':` part? Potentially make `verb = ""` and see if it set before you wait for the task and mark as changed?

**Review by pdellaert — CHANGES_REQUESTED** [review]

File: lib/ansible/modules/cloud/vmware/vmware_host_powerstate.py (hunk 3 above)
Comment (Akasurde) [reply lib/ansible/modules/cloud/vmware/vmware_host_powerstate.py:42]: Done.

File: lib/ansible/modules/cloud/vmware/vmware_host_powerstate.py (hunk 1 above)
Comment (Akasurde) [reply lib/ansible/modules/cloud/vmware/vmware_host_powerstate.py]: OK.

File: lib/ansible/modules/cloud/vmware/vmware_host_powerstate.py (hunk 4 above)
Comment (Akasurde) [reply lib/ansible/modules/cloud/vmware/vmware_host_powerstate.py:146]: Need to try.

File: lib/ansible/modules/cloud/vmware/vmware_host_powerstate.py (hunk 6 above)
Comment (Akasurde) [reply lib/ansible/modules/cloud/vmware/vmware_host_powerstate.py]: Done.

File: lib/ansible/modules/cloud/vmware/vmware_host_powerstate.py (hunk 5 above)
Comment (Akasurde) [reply lib/ansible/modules/cloud/vmware/vmware_host_powerstate.py]: Done.

File: lib/ansible/modules/cloud/vmware/vmware_host_powerstate.py (hunk 7 above)
Comment (Akasurde) [reply lib/ansible/modules/cloud/vmware/vmware_host_powerstate.py]: Done.

File: lib/ansible/modules/cloud/vmware/vmware_host_powerstate.py (hunk 4 above)
Comment (Akasurde) [reply lib/ansible/modules/cloud/vmware/vmware_host_powerstate.py:146]: I am afraid that my setup does not provide facility to set host in standby mode.

**Review by pdellaert — APPROVED** [review]
LGTM

**Comment by Akasurde** [comment]
@pdellaert Thanks for the review.

**Pull request merged**